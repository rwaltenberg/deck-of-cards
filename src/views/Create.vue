<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Deck from '@/utils/deck'

@Component
export default class Create extends Vue {
  cards = (new Array(10)).fill('').map((c, i) => ({ index: i, value: '' }))
  rotation = { value: '' }
  loading = false

  async save () {
    const cards = this.cards.filter(card => card.value).map(card => card.value)
    const rotation = this.rotation.value

    if (!cards.length || !rotation || this.loading) {
      return
    }

    this.loading = true
    const deck = new Deck(cards, rotation)

    try {
      await deck.save()
      this.$router.push({ name: 'Deck', params: { id: deck.getId() || '' } })
    } catch (error) {
      console.error(error)
      this.loading = false
    }
  }
}
</script>

<template>
  <form class="create" @submit.prevent="save">
    <div class="cards">
      <div class="card" v-for="card in cards" :key="`card${card.index}`">
        <label :for="`card${card.index}`">Card {{ card.index + 1 }}</label>
        <input :id="`card${card.index}`" type="text" maxlength="2" v-model="card.value" placeholder="Enter card">
      </div>
    </div>
    <div class="rotation">
      <div class="card">
        <label for="rotation">Rotation Card</label>
        <input id="rotation" type="text" maxlength="2" v-model="rotation.value" placeholder="Enter card">
      </div>
    </div>
    <button class="submit" type="submit">Submit</button>
  </form>
</template>

<style lang="scss" scoped>
.create {
  align-items: center;
  display: flex;
  flex-direction: column;
  margin: 50px auto;
  max-width: 1280px;
}

.cards {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;

  .card {
    flex: 0 0 calc(20% - 5px);
    margin-bottom: rem(25px);

    label {
      color: #AAA;
      display: block;
      font-size: rem(10px);
      font-weight: bold;
      margin-bottom: rem(5px);
      text-transform: uppercase;
    }

    input {
      display: block;
      width: 100%;
    }
  }
}

.card {
  input {
    border: 1px solid #CCC;
    border-radius: 5px;
    box-shadow: 0 0 5px rgba(#CCC, .5);
    box-sizing: border-box;
    display: inline-block;
    font-size: rem(14px);
    line-height: 1;
    padding: rem(7px) rem(10px);
    text-transform: uppercase;

    &::placeholder {
      text-transform: none;
    }
  }
}

.rotation {
  margin: 50px 0 300px;
  width: calc(60% - 5px);

  .card {
    label {
      box-sizing: border-box;
      display: inline-block;
      font-size: rem(24px);
      font-weight: bold;
      margin-right: 3px;
      padding-right: 10px;
      text-align: right;
      vertical-align: bottom;
      width: calc(100% / 3 - 5px);
    }

    input {
      margin: 0 5px;
      width: calc(100% / 3 - 5px);
    }
  }
}

.submit {
  background-color: #190057;
  border: 0;
  border-radius: 15px;
  color: #f1e622;
  font-family: Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-size: rem(24px);
  font-weight: bold;
  padding: 8px 30px;
}
</style>
