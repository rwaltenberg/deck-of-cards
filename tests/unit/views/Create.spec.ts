import Vuex, { Store } from 'vuex'
import { Wrapper, createLocalVue, shallowMount } from '@vue/test-utils'

import Create from '@/views/Create.vue'

function setInputValue (input: Wrapper<Create>, value: string) {
  const el = input.element as HTMLInputElement
  el.value = value
  input.trigger('input')
}

describe('Create.vue', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(Create)
    expect(wrapper).toMatchSnapshot()
  })

  describe('Form submition', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    let actions!: { createDeck: jest.Mock }
    let store!: Store<any>

    beforeEach(() => {
      actions = { createDeck: jest.fn(async () => {}) }
      store = new Vuex.Store({ actions })
    })

    it('does not trigger "save" until there is at least one card and one rotation card', () => {
      const wrapper = shallowMount(Create, {
        store,
        localVue,
        mocks: {
          $router: { push: () => {} }
        }
      })

      const card = wrapper.find('#card0')
      const rotation = wrapper.find('#rotation')
      const form = wrapper.find('form').element as HTMLFormElement

      form.submit()
      expect(actions.createDeck).not.toBeCalled()

      setInputValue(card, 'AS')
      card.trigger('input')

      form.submit()
      expect(actions.createDeck).not.toBeCalled()

      setInputValue(rotation, 'AS')
      rotation.trigger('input')

      form.submit()
      expect(actions.createDeck).toBeCalled()

      expect(actions.createDeck).toHaveBeenCalledTimes(1)
    })

    it('triggers "save" only once until finished loading', () => {
      const wrapper = shallowMount(Create, {
        store,
        localVue,
        mocks: {
          $router: { push: () => {} }
        }
      })

      const card = wrapper.find('#card0')
      const rotation = wrapper.find('#rotation')
      const form = wrapper.find('form').element as HTMLFormElement

      setInputValue(card, 'AS')
      card.trigger('input')

      setInputValue(rotation, 'AS')
      rotation.trigger('input')

      form.submit()
      form.submit()

      expect(actions.createDeck).toHaveBeenCalledTimes(1)
    })
  })
})
