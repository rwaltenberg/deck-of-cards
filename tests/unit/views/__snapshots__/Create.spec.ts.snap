// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create.vue renders correctly 1`] = `
<form class="create">
  <div class="cards">
    <div class="card"><label for="card0">Card 1</label> <input id="card0" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card1">Card 2</label> <input id="card1" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card2">Card 3</label> <input id="card2" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card3">Card 4</label> <input id="card3" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card4">Card 5</label> <input id="card4" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card5">Card 6</label> <input id="card5" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card6">Card 7</label> <input id="card6" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card7">Card 8</label> <input id="card7" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card8">Card 9</label> <input id="card8" type="text" maxlength="2" placeholder="Enter card"></div>
    <div class="card"><label for="card9">Card 10</label> <input id="card9" type="text" maxlength="2" placeholder="Enter card"></div>
  </div>
  <div class="rotation">
    <div class="card"><label for="rotation">Rotation Card</label> <input id="rotation" type="text" maxlength="2" placeholder="Enter card"></div>
  </div> <button type="submit" class="submit">Submit</button>
</form>
`;
